src/main/java/org/mockito/internal/creation/bytebuddy/CachingMockBytecodeGenerator.java

```diff
@@ -1,70 +1,94 @@
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.locks.Lock;
-import java.util.concurrent.locks.ReentrantLock;
-import org.mockito.exceptions.base.MockitoException;
+import java.util.concurrent.ConcurrentMap;
+
+import static org.mockito.internal.util.StringJoiner.join;
 
-class CachingMockBytecodeGenerator {
+class CachingMockBytecodeGenerator extends ReferenceQueue<ClassLoader> {
 
-    private final Lock avoidingClassLeakCacheLock = new ReentrantLock();
-    public final WeakHashMap<ClassLoader, CachedBytecodeGenerator> avoidingClassLeakageCache =
-            new WeakHashMap<ClassLoader, CachedBytecodeGenerator>();
+    private static final ClassLoader BOOT_LOADER = new URLClassLoader(new URL[0], null);
+
+    final ConcurrentMap<Key<ClassLoader>, CachedBytecodeGenerator> avoidingClassLeakageCache =
+            new ConcurrentHashMap<Key<ClassLoader>, CachedBytecodeGenerator>();
 
     private final MockBytecodeGenerator mockBytecodeGenerator = new MockBytecodeGenerator();
 
+    private final boolean weak;
+
+    public CachingMockBytecodeGenerator(boolean weak) {
+        this.weak = weak;
+    }
+
     @SuppressWarnings("unchecked")
     public <T> Class<T> get(MockFeatures<T> params) {
-        // TODO improves locking behavior with ReentrantReadWriteLock ?
-        avoidingClassLeakCacheLock.lock();
-        try {
-
-            Class<?> generatedMockClass = mockCachePerClassLoaderOf(params.mockedType).getOrGenerateMockClass(
-                    params
-            );
+        pollAndClearReferences();
+        return (Class<T>) mockCachePerClassLoaderOf(params.mockedType.getClassLoader()).getOrGenerateMockClass(params);
+    }
 
-            return (Class<T>) generatedMockClass;
-        } finally {
-          avoidingClassLeakCacheLock.unlock();
+    void pollAndClearReferences() {
+        Reference<?> reference;
+        while ((reference = poll()) != null) {
+            avoidingClassLeakageCache.remove(reference);
         }
     }
 
-    private <T> CachedBytecodeGenerator mockCachePerClassLoaderOf(Class<T> mockedType) {
-        if (!avoidingClassLeakageCache.containsKey(mockedType.getClassLoader())) {
-            avoidingClassLeakageCache.put(
-                    mockedType.getClassLoader(),
-                    new CachedBytecodeGenerator(mockBytecodeGenerator)
-            );
+    private CachedBytecodeGenerator mockCachePerClassLoaderOf(ClassLoader classLoader) {
+        classLoader = classLoader == null ? BOOT_LOADER : classLoader;
+        CachedBytecodeGenerator generator = avoidingClassLeakageCache.get(new StrongKey<ClassLoader>(classLoader)), newGenerator;
+        if (generator == null) {
+            generator = avoidingClassLeakageCache.putIfAbsent(new WeakKey<ClassLoader>(classLoader, this), newGenerator = new CachedBytecodeGenerator(mockBytecodeGenerator, weak));
+            if (generator == null) {
+                generator = newGenerator;
+            }
```

Comment:
I'd change that code to one variable declaration per line. And I would avoid inline assignment too (the first `newGenerator` is not needed before the `if`).
Those too may confuse most readers.